Error while checking build/O2/fullCI for c5c77e34381838589a83d72b6047375ecf2bb1c4 at 2023-04-04 15:06:
```
## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes
[6169:BadMapCalibSpec]: [12:36:25][ERROR] Insufficient statistics: 5 entries in lowE histo, do nothing


## sw/BUILD/o2checkcode-latest/log
--
++ echo '========== List of errors found =========='
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/b8d65e40231dca94933325f6be7665d99ed31ac3/slc8_x86-64/o2checkcode/1.0-local2/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11073/c5c77e34381838589a83d72b6047375ecf2bb1c4/build_O2_fullCI/pretty.html).

--- a/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/RecoContainer.h
+++ b/Detectors/EMCAL/reconstruction/include/EMCALReconstruction/RecoContainer.h
@@ -29,6 +29,15 @@ namespace o2::emcal
 {
 /// \struct RecCellInfo
 /// \brief Container class for cell information for merging
+///
+/// In case the energy is in the overlap region between the
+/// two digitizers 2 channels exist for the same cell. In this
+/// case the low gain cells are used above a certain threshold.
+/// In certain error cases the information from the other digitizer
+/// might be missing. Such cases must be fitered out, however this
+/// can be done only after all cells are processed. The overlap information
+/// needs to be propagated for the filtering but is not part of the
+/// final cell object
 struct RecCellInfo {
   o2::emcal::Cell mCellData; ///< Cell information
   bool mIsLGnoHG;            ///< Cell has only LG digits
@@ -144,7 +153,7 @@ class EventContainer
   bool isCellSaturated(double energy) const;
 
   o2::InteractionRecord mInteractionRecord;
-  uint64_t mTriggerBits;             ///< Trigger bits of the event
+  uint64_t mTriggerBits = 0;         ///< Trigger bits of the event
   std::vector<RecCellInfo> mCells;   ///< Container of cells in event
   std::vector<RecCellInfo> mLEDMons; ///< Container of LEDMONs in event
 };

--- a/Detectors/EMCAL/workflow/include/EMCALWorkflow/RawToCellConverterSpec.h
+++ b/Detectors/EMCAL/workflow/include/EMCALWorkflow/RawToCellConverterSpec.h
@@ -10,6 +10,7 @@
 // or submit itself to any jurisdiction.
 
 #include <chrono>
+#include <exception>
 #include <vector>
 
 #include "Framework/ConcreteDataMatcher.h"
@@ -22,6 +23,7 @@
 #include "EMCALBase/Mapper.h"
 #include "EMCALReconstruction/CaloRawFitter.h"
 #include "EMCALReconstruction/RecoContainer.h"
+#include "EMCALReconstruction/ReconstructionErrors.h"
 
 namespace o2
 {
@@ -94,27 +96,88 @@ class RawToCellConverterSpec : public framework::Task
   header::DataHeader::SubSpecificationType getSubspecification() const { return mSubspecification; }
 
  private:
-  /// \struct RecCellInfo
-  /// \brief Internal bookkeeping for cell double counting
-  ///
-  /// In case the energy is in the overlap region between the
-  /// two digitizers 2 channels exist for the same cell. In this
-  /// case the low gain cells are used above a certain threshold.
-  /// In certain error cases the information from the other digitizer
-  /// might be missing. Such cases must be fitered out, however this
-  /// can be done only after all cells are processed. The overlap information
-  /// needs to be propagated for the filtering but is not part of the
-  /// final cell object
-  struct RecCellInfo {
-    o2::emcal::Cell mCellData; ///< Cell information
-    bool mIsLGnoHG;            ///< Cell has only LG digits
-    bool mHGOutOfRange;        ///< Cell has only HG digits which are out of range
-    int mFecID;                ///< FEC ID of the channel (for monitoring)
-    int mDDLID;                ///< DDL of the channel (for monitoring)
-    int mHWAddressLG;          ///< HW address of LG (for monitoring)
-    int mHWAddressHG;          ///< HW address of HG (for monitoring)
+  /// \class ModuleIndexException
+  /// \brief Exception handling errors in calculation of the absolute module ID
+  class ModuleIndexException : public std::exception
+  {
+   public:
+    /// \enum ModuleType_t
+    /// \brief Type of module raising the exception
+    enum class ModuleType_t {
+      CELL_MODULE,  ///< Cell module type
+      LEDMON_MODULE ///< LEDMON module type
+    };
+
+    /// \brief Constructor for cell indices
+    /// \param moduleIndex Index of the module raising the exception
+    /// \param column Column of the cell
+    /// \param row Row of the cell
+    /// \param columnshifted Shifted column index
+    /// \param rowshifted Shifted row index
+    ModuleIndexException(int moduleIndex, int column, int row, int columnshifted, int rowshifted);
+
+    /// \brief Constructor for LEDMON indices
+    /// \param moduleIndex Index of the module raising the exception
+    ModuleIndexException(int moduleIndex);
+
+    /// \brief Destructor
+    ~ModuleIndexException() noexcept final = default;
+
+    /// \brief Access to error message
+    /// \return Error message
+    const char* what() const noexcept final { return "Invalid cell / LEDMON index"; }
+
+    /// \brief Get type of module raising the exception
+    /// \return Module type
+    ModuleType_t getModuleType() const { return mModuleType; }
+
+    /// \brief Get index of the module raising the exception
+    /// \return Index of the module
+    int getIndex() const { return mIndex; }
+
+    /// \brief Get column raising the exception (cell-case)
+    /// \return Column
+    int getColumn() const { return mColumn; }
+
+    /// \brief Get row raising the exception (cell-case)
+    /// \return Row
+    int getRow() const { return mRow; }
+
+    /// \brief Get shifted column raising the exception (cell-case)
+    /// \return Shifted column
+    int getColumnShifted() const { return mColumnShifted; }
+
+    /// \brief Get shifted row raising the exception (cell-case)
+    /// \return Shifted row
+    int getRowShifted() const { return mRowShifted; }
+
+   private:
+    ModuleType_t mModuleType; ///< Type of the module raising the exception
+    int mIndex = -1;          ///< Index raising the exception
+    int mColumn = -1;         ///< Column of the module (cell-case)
+    int mRow = -1;            ///< Row of the module (cell-case)
+    int mColumnShifted = -1;  ///< shifted column of the module (cell-case)
+    int mRowShifted = -1;     /// << shifted row of the module (cell-case)
   };
+
+  /// \brief Check if the timeframe is empty
+  /// \param ctx Processing context of timeframe
+  /// \return True if the timeframe is empty, false otherwise
+  ///
+  /// Emtpy timeframes do not have RAWDATA from any physical link in the
+  /// processing context, instead they contain RAWDATA from link 0xDEADBEEF
+  /// and a message in FLP/DISTSUBTIMEFRAME
   bool isLostTimeframe(framework::ProcessingContext& ctx) const;
+
+  /// \brief Select cells and put them on the output container
+  /// \param cells Cells to select
+  /// \param isLEDMON Distinction whether input is cell or LEDMON
+  /// \return Number of accepted cells
+  ///
+  /// Cells or LEDMONs are rejected if
+  /// - They have a low gain but no high gain channel
+  /// - They only have a high gain channel which is out of range
+  int bookEventCells(const gsl::span<const o2::emcal::RecCellInfo>& cells, bool isLELDMON);
 
   /// \brief Send data to output channels
   /// \param cells Container with output cells for timeframe
@@ -125,6 +188,28 @@ class RawToCellConverterSpec : public framework::Task
   /// is determined on the fly in the run method and therefore used as parameter. Consumers
   /// must use wildcard subspecification via ConcreteDataTypeMatcher.
   void sendData(framework::ProcessingContext& ctx, const std::vector<o2::emcal::Cell>& cells, const std::vector<o2::emcal::TriggerRecord>& triggers, const std::vector<ErrorTypeFEE>& decodingErrors) const;
+
+  /// \brief Get absolute Cell ID from column/row in supermodule
+  /// \param supermoduleID Index of the supermodule
+  /// \param column Column of the tower within the supermodule
+  /// \param row Row of the tower within the supermodule
+  /// \return Cell absolute ID
+  /// \throw ModuleIndexException in case of invalid module indices
+  int getCellAbsID(int supermoduleID, int column, int row);
+
+  /// \brief Get the absoulte ID of LEDMON from the module ID in supermodule
+  /// \param supermoduleID Index of the supermodule
+  /// \param module Index of the module within the supermodule
+  /// \return LEDMON absolute ID
+  /// \throw ModuleIndexException in case of invalid module indices
+  int geLEDMONAbsID(int supermoduleID, int module);
+
+  /// \brief handler function for gain type errors
+  /// \param errortype Gain error type
+  /// \param ddlID ID of the DDL
+  /// \param fecID ID of the FEC
+  /// \param hwaddress Hardware address
+  void handleGainError(o2::emcal::reconstructionerrors::GainError_t errortype, int ddlID, int fecID, int hwaddress);
 
   header::DataHeader::SubSpecificationType mSubspecification = 0;    ///< Subspecification for output channels
   int mNoiseThreshold = 0;                                           ///< Noise threshold in raw fit

--- a/Detectors/EMCAL/workflow/src/RawToCellConverterSpec.cxx
+++ b/Detectors/EMCAL/workflow/src/RawToCellConverterSpec.cxx
@@ -38,7 +38,6 @@
 #include "EMCALReconstruction/AltroDecoder.h"
 #include "EMCALReconstruction/RawDecodingError.h"
 #include "EMCALReconstruction/RecoParam.h"
-#include "EMCALReconstruction/ReconstructionErrors.h"
 #include "EMCALWorkflow/RawToCellConverterSpec.h"
 #include "SimulationDataFormat/MCCompLabel.h"
 #include "SimulationDataFormat/MCTruthContainer.h"
@@ -111,7 +110,6 @@ void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
   double timeshift = RecoParam::Instance().getCellTimeShiftNanoSec(); // subtract offset in ns in order to center the time peak around the nominal delay
   constexpr auto originEMC = o2::header::gDataOriginEMC;
   constexpr auto descRaw = o2::header::gDataDescriptionRawData;
-  double noiseThresholLGnoHG = RecoParam::Instance().getNoiseThresholdLGnoHG();
 
   // reset message counter after 10 minutes
   auto currenttime = std::chrono::system_clock::now();
@@ -343,43 +341,27 @@ void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
             continue;
           }
 
-          if (!(chantype == o2::emcal::ChannelType_t::HIGH_GAIN || chantype == o2::emcal::ChannelType_t::LOW_GAIN)) {
+          if (!(chantype == o2::emcal::ChannelType_t::HIGH_GAIN || chantype == o2::emcal::ChannelType_t::LOW_GAIN || chantype == o2::emcal::ChannelType_t::LEDMON)) {
             continue;
           }
 
-          auto [phishift, etashift] = mGeometry->ShiftOnlineToOfflineCellIndexes(iSM, iRow, iCol);
-          int CellID = mGeometry->GetAbsCellIdFromCellIndexes(iSM, phishift, etashift);
-          if (CellID > 17664) {
-            if (mNumErrorMessages < mMaxErrorMessages) {
-              std::string celltypename;
-              switch (chantype) {
-                case o2::emcal::ChannelType_t::HIGH_GAIN:
-                  celltypename = "high gain";
-                  break;
-                case o2::emcal::ChannelType_t::LOW_GAIN:
-                  celltypename = "low-gain";
-                  break;
-                case o2::emcal::ChannelType_t::TRU:
-                  celltypename = "TRU";
-                  break;
-                case o2::emcal::ChannelType_t::LEDMON:
-                  celltypename = "LEDMON";
-                  break;
-              };
-              LOG(warning) << "Sending invalid cell ID " << CellID << "(SM " << iSM << ", row " << iRow << " - shift " << phishift << ", col " << iCol << " - shift " << etashift << ") of type " << celltypename;
-              mNumErrorMessages++;
-              if (mNumErrorMessages == mMaxErrorMessages) {
-                LOG(warning) << "Max. amount of error messages (" << mMaxErrorMessages << " reached, further messages will be suppressed";
-              }
-            } else {
-              mErrorMessagesSuppressed++;
-            }
-            if (mCreateRawDataErrors) {
-              mOutputDecoderErrors.emplace_back(feeID, ErrorTypeFEE::ErrorSource_t::GEOMETRY_ERROR, reconstructionerrors::getErrorCodeFromGeometryError(reconstructionerrors::GeometryError_t::CELL_RANGE_EXCEED), CellID, chan.getHardwareAddress()); // 0 -> Cell ID out of range
-            }
+          // Drop LEDMON reconstruction in case of physics triggers
+          if (chantype == o2::emcal::ChannelType_t::LEDMON && !(triggerbits & o2::trigger::Cal)) {
             continue;
           }
-          if (CellID < 0) {
+
+          int CellID = -1;
+          bool isLowGain = false;
+          try {
+            if (chantype == o2::emcal::ChannelType_t::HIGH_GAIN || chantype == o2::emcal::ChannelType_t::LOW_GAIN) {
+              // high- / low-gain cell
+              CellID = getCellAbsID(iSM, iCol, iRow);
+              isLowGain = chantype == o2::emcal::ChannelType_t::LOW_GAIN;
+            } else {
+              CellID = geLEDMONAbsID(iSM, iCol); // Module index encoded in colum for LEDMONs
+              isLowGain = iRow == 0;             // For LEDMONs gain type is encoded in the row (0 - low gain, 1 - high gain)
+            }
+          } catch (ModuleIndexException& e) {
             if (mNumErrorMessages < mMaxErrorMessages) {
               std::string celltypename;
               switch (chantype) {
@@ -396,7 +378,14 @@ void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
                   celltypename = "LEDMON";
                   break;
               };
-              LOG(warning) << "Sending negative cell ID " << CellID << "(SM " << iSM << ", row " << iRow << " - shift " << phishift << ", col " << iCol << " - shift " << etashift << ") of type " << celltypename;
+              switch (e.getModuleType()) {
+                case ModuleIndexException::ModuleType_t::CELL_MODULE:
+                  LOG(warning) << "Sending invalid or negative cell ID " << e.getIndex() << " (SM " << iSM << ", row " << e.getRow() << " - shift " << e.getRowShifted() << ", col " << e.getColumn() << " - shift " << e.getColumnShifted() << ") of type " << celltypename;
+                  break;
+                case ModuleIndexException::ModuleType_t::LEDMON_MODULE:
+                  LOG(warning) << "Sending invalid or negative LEDMON module ID " << e.getIndex() << "( SM" << iSM << ")";
+                  break;
+              };
               mNumErrorMessages++;
               if (mNumErrorMessages == mMaxErrorMessages) {
                 LOG(warning) << "Max. amount of error messages (" << mMaxErrorMessages << " reached, further messages will be suppressed";
@@ -405,7 +394,7 @@ void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
               mErrorMessagesSuppressed++;
             }
             if (mCreateRawDataErrors) {
-              mOutputDecoderErrors.emplace_back(feeID, ErrorTypeFEE::ErrorSource_t::GEOMETRY_ERROR, reconstructionerrors::getErrorCodeFromGeometryError(reconstructionerrors::GeometryError_t::CELL_INDEX_NEGATIVE), CellID, chan.getHardwareAddress()); // Geometry error codes will start from 100
+              mOutputDecoderErrors.emplace_back(feeID, ErrorTypeFEE::ErrorSource_t::GEOMETRY_ERROR, reconstructionerrors::getErrorCodeFromGeometryError(CellID < 0 ? reconstructionerrors::GeometryError_t::CELL_INDEX_NEGATIVE : reconstructionerrors::GeometryError_t::CELL_RANGE_EXCEED), CellID, chan.getHardwareAddress()); // 0 -> Cell ID out of range
             }
             continue;
           }
@@ -424,11 +413,16 @@ void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
             // apply correction for bc mod 4
             double celltime = fitResults.getTime() - timeshift - 25 * bcmod4;
             double amp = fitResults.getAmp() * o2::emcal::constants::EMCAL_ADCENERGY;
-            if (chantype == ChannelType_t::LOW_GAIN) {
+            if (isLowGain) {
               amp *= o2::emcal::constants::EMCAL_HGLGFACTOR;
             }
             int fecID = mMapper->getFEEForChannelInDDL(feeID, chan.getFECIndex(), chan.getBranchIndex());
-            currentEvent.setCell(CellID, amp, celltime, chantype, chan.getHardwareAddress(), fecID, feeID, mMergeLGHG);
+            if (chantype == o2::emcal::ChannelType_t::LEDMON) {
+              // Mark LEDMONs as HIGH_GAIN/LOW_GAIN for gain type merging - will be flagged as LEDMON later when pushing to the output container
+              currentEvent.setLEDMONCell(CellID, amp, celltime, isLowGain ? o2::emcal::ChannelType_t::LOW_GAIN : o2::emcal::ChannelType_t::HIGH_GAIN, chan.getHardwareAddress(), fecID, feeID, mMergeLGHG);
+            } else {
+              currentEvent.setCell(CellID, amp, celltime, chantype, chan.getHardwareAddress(), fecID, feeID, mMergeLGHG);
+            }
           } catch (CaloRawFitter::RawFitterError_t& fiterror) {
             if (fiterror != CaloRawFitter::RawFitterError_t::BUNCH_NOT_OK) {
               // Display
@@ -470,57 +464,24 @@ void RawToCellConverterSpec::run(framework::ProcessingContext& ctx)
   // Loop over BCs, sort cells with increasing tower ID and write to output containers
   RecoContainerReader eventIterator(mCellHandler);
   while (eventIterator.hasNext()) {
-    int ncellsEvent = 0;
+    int ncellsEvent = 0, nLEDMONsEvent = 0;
     int eventstart = mOutputCells.size();
     auto& currentevent = eventIterator.nextEvent();
+    // Add cells
     if (currentevent.getNumberOfCells()) {
       LOG(debug) << "Event has " << currentevent.getNumberOfCells() << " cells";
       currentevent.sortCells(false);
-      for (const auto& cell : currentevent.getCells()) {
-        if (cell.mIsLGnoHG) {
-          // Treat error only in case the LG is above the noise threshold
-          // no HG cell found, we can assume the cell amplitude is the LG amplitude
-          int ampLG = cell.mCellData.getAmplitude() / (o2::emcal::constants::EMCAL_ADCENERGY * o2::emcal::constants::EMCAL_HGLGFACTOR);
-          // use cut at 3 sigma where sigma for the LG digitizer is 0.4 ADC counts (EMCAL-502)
-          if (ampLG > noiseThresholLGnoHG) {
-            if (mNumErrorMessages < mMaxErrorMessages) {
-              LOG(warning) << "FEC " << cell.mFecID << ": 0x" << std::hex << cell.mHWAddressLG << std::dec << " (DDL " << cell.mDDLID << ") has low gain but no high-gain";
-              mNumErrorMessages++;
-              if (mNumErrorMessages == mMaxErrorMessages) {
-                LOG(warning) << "Max. amount of error messages (" << mMaxErrorMessages << " reached, further messages will be suppressed";
-              }
-            } else {
-              mErrorMessagesSuppressed++;
-            }
-            if (mCreateRawDataErrors) {
-              mOutputDecoderErrors.emplace_back(cell.mDDLID, ErrorTypeFEE::GAIN_ERROR, reconstructionerrors::getErrorCodeFromGainError(reconstructionerrors::GainError_t::LGNOHG), cell.mFecID, cell.mHWAddressLG);
-            }
-          }
-          continue;
-        }
-        if (cell.mHGOutOfRange) {
-          if (mNumErrorMessages < mMaxErrorMessages) {
-            LOG(warning) << "FEC " << cell.mFecID << ": 0x" << std::hex << cell.mHWAddressHG << std::dec << " (DDL " << cell.mDDLID << ") has only high-gain out-of-range";
-            mNumErrorMessages++;
-            if (mNumErrorMessages == mMaxErrorMessages) {
-              LOG(warning) << "Max. amount of error messages (" << mMaxErrorMessages << " reached, further messages will be suppressed";
-            }
-          } else {
-            mErrorMessagesSuppressed++;
-          }
-          if (mCreateRawDataErrors) {
-            mOutputDecoderErrors.emplace_back(cell.mDDLID, ErrorTypeFEE::GAIN_ERROR, reconstructionerrors::getErrorCodeFromGainError(reconstructionerrors::GainError_t::HGNOLG), cell.mFecID, cell.mHWAddressHG);
-          }
-          continue;
-        }
-        ncellsEvent++;
-        mOutputCells.push_back(cell.mCellData);
-      }
-
-      const auto interaction = currentevent.getInteractionRecord();
-      LOG(debug) << "Next event [Orbit " << interaction.orbit << ", BC (" << interaction.bc << "]: Accepted " << ncellsEvent << " cells";
-      mOutputTriggerRecords.emplace_back(interaction, currentevent.getTriggerBits(), eventstart, ncellsEvent);
+      ncellsEvent = bookEventCells(currentevent.getCells(), false);
     }
+    // Add LEDMONs (if present)
+    if (currentevent.getNumberOfLEDMONs()) {
+      LOG(debug) << "Event has " << currentevent.getNumberOfLEDMONs() << " LEDMONs";
+      currentevent.sortCells(true);
+      nLEDMONsEvent = bookEventCells(currentevent.getLEDMons(), true);
+    }
+    const auto interaction = currentevent.getInteractionRecord();
+    LOG(debug) << "Next event [Orbit " << interaction.orbit << ", BC (" << interaction.bc << "]: Accepted " << ncellsEvent << " cells and " << nLEDMONsEvent << " LEDMONS";
+    mOutputTriggerRecords.emplace_back(interaction, currentevent.getTriggerBits(), eventstart, ncellsEvent + nLEDMONsEvent);
   }
 
   LOG(info) << "[EMCALRawToCellConverter - run] Writing " << mOutputCells.size() << " cells from " << mOutputTriggerRecords.size() << " events ...";
@@ -562,6 +523,77 @@ bool RawToCellConverterSpec::isLostTimeframe(framework::ProcessingContext& ctx) 
   return false;
 }
 
+int RawToCellConverterSpec::bookEventCells(const gsl::span<const o2::emcal::RecCellInfo>& cells, bool isLELDMON)
+{
+  double noiseThresholLGnoHG = RecoParam::Instance().getNoiseThresholdLGnoHG();
+  int ncellsSelected = 0;
+  for (const auto& cell : cells) {
+    if (cell.mIsLGnoHG) {
+      // Treat error only in case the LG is above the noise threshold
+      // no HG cell found, we can assume the cell amplitude is the LG amplitude
+      int ampLG = cell.mCellData.getAmplitude() / (o2::emcal::constants::EMCAL_ADCENERGY * o2::emcal::constants::EMCAL_HGLGFACTOR);
+      // use cut at 3 sigma where sigma for the LG digitizer is 0.4 ADC counts (EMCAL-502)
+      if (ampLG > noiseThresholLGnoHG) {
+        handleGainError(reconstructionerrors::GainError_t::LGNOHG, cell.mDDLID, cell.mFecID, cell.mHWAddressLG);
+      }
+      continue;
+    }
+    if (cell.mHGOutOfRange) {
+      handleGainError(reconstructionerrors::GainError_t::HGNOLG, cell.mDDLID, cell.mFecID, cell.mHWAddressHG);
+      continue;
+    }
+    ncellsSelected++;
+    mOutputCells.push_back(cell.mCellData);
+    if (isLELDMON) {
+      // LEDMON was handled internally in the reco container as HG/LG cell for gain type merging - reflag as LEDMON
+      mOutputCells.back().setLEDMon();
+    }
+  }
+  return ncellsSelected;
+}
+
+int RawToCellConverterSpec::getCellAbsID(int supermoduleID, int column, int row)
+{
+  auto [phishift, etashift] = mGeometry->ShiftOnlineToOfflineCellIndexes(supermoduleID, row, column);
+  int cellID = mGeometry->GetAbsCellIdFromCellIndexes(supermoduleID, phishift, etashift);
+  if (cellID > 17664 || cellID < 0) {
+    throw ModuleIndexException(cellID, column, row, etashift, phishift);
+  }
+  return cellID;
+}
+
+int RawToCellConverterSpec::geLEDMONAbsID(int supermoduleID, int moduleID)
+{
+  if (moduleID >= o2::emcal::EMCAL_LEDREFS || moduleID < 0) {
+    throw ModuleIndexException(moduleID);
+  }
+  return supermoduleID * o2::emcal::EMCAL_LEDREFS + moduleID;
+}
+
+void RawToCellConverterSpec::handleGainError(o2::emcal::reconstructionerrors::GainError_t errortype, int ddlID, int fecID, int hwaddress)
+{
+  if (mNumErrorMessages < mMaxErrorMessages) {
+    switch (errortype) {
+      case reconstructionerrors::GainError_t::HGNOLG:
+        LOG(warning) << "FEC " << fecID << ": 0x" << std::hex << hwaddress << std::dec << " (DDL " << ddlID << ") has only high-gain out-of-range";
+        break;
+      case reconstructionerrors::GainError_t::LGNOHG:
+        LOG(warning) << "FEC " << fecID << ": 0x" << std::hex << hwaddress << std::dec << " (DDL " << ddlID << ") has low gain but no high-gain";
+        break;
+      default:
+        LOG(warning) << "FEC " << fecID << ": 0x" << std::hex << hwaddress << std::dec << " (DDL " << ddlID << ") falsely flagged as gain error";
+    };
+    if (mNumErrorMessages == mMaxErrorMessages) {
+      LOG(warning) << "Max. amount of error messages (" << mMaxErrorMessages << " reached, further messages will be suppressed";
+    }
+  } else {
+    mErrorMessagesSuppressed++;
+  }
+  if (mCreateRawDataErrors) {
+    mOutputDecoderErrors.emplace_back(ddlID, ErrorTypeFEE::GAIN_ERROR, reconstructionerrors::getErrorCodeFromGainError(errortype), fecID, hwaddress);
+  }
+}
+
 void RawToCellConverterSpec::sendData(framework::ProcessingContext& ctx, const std::vector<o2::emcal::Cell>& cells, const std::vector<o2::emcal::TriggerRecord>& triggers, const std::vector<ErrorTypeFEE>& decodingErrors) const
 {
   constexpr auto originEMC = o2::header::gDataOriginEMC;
@@ -572,6 +604,15 @@ void RawToCellConverterSpec::sendData(framework::ProcessingContext& ctx, const s
     ctx.outputs().snapshot(framework::Output{originEMC, "DECODERERR", mSubspecification, framework::Lifetime::Timeframe}, decodingErrors);
   }
 }
+
+RawToCellConverterSpec::ModuleIndexException::ModuleIndexException(int moduleIndex, int column, int row, int shiftedColumn, int shiftedRow) : mModuleType(ModuleType_t::CELL_MODULE),
+                                                                                                                                              mIndex(moduleIndex),
+                                                                                                                                              mColumn(column),
+                                                                                                                                              mRow(row),
+                                                                                                                                              mColumnShifted(shiftedColumn),
+                                                                                                                                              mRowShifted(shiftedRow) {}
+
+RawToCellConverterSpec::ModuleIndexException::ModuleIndexException(int moduleIndex) : mModuleType(ModuleType_t::LEDMON_MODULE), mIndex(moduleIndex) {}
 
 o2::framework::DataProcessorSpec o2::emcal::reco_workflow::getRawToCellConverterSpec(bool askDISTSTF, bool disableDecodingErrors, int subspecification)
 {

